File: src/main/java/com/softwareplace/security/authorization/JWTAuthorizationFilter.java
package com.softwareplace.security.authorization;

import static org.springframework.http.HttpHeaders.AUTHORIZATION;

import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

import javax.servlet.FilterChain;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;
import org.springframework.web.client.HttpClientErrorException;

import com.softwareplace.security.model.UserData;
import com.softwareplace.security.service.AuthorizationUserService;
import com.softwareplace.security.util.Constants;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.SignatureException;

public class JWTAuthorizationFilter extends BasicAuthenticationFilter {

	private final AuthorizationUserService authorizationUserService;
	private final AuthorizationHandler authorizationHandler;

	public JWTAuthorizationFilter(AuthenticationManager authenticationManager,
			AuthorizationUserService authorizationUserService, AuthorizationHandler authorizationHandler) {
		super(authenticationManager);
		this.authorizationUserService = authorizationUserService;
		this.authorizationHandler = authorizationHandler;
	}

	@Override protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
			throws IOException {
		if (shouldBeAuthorized(request)) {
			try {
				SecurityContextHolder.getContext().setAuthentication(getUsernamePasswordAuthenticationToken(request));
				chain.doFilter(request, response);
			} catch (HttpClientErrorException.Unauthorized | MalformedJwtException | SignatureException exception) {
				response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
				ResponseRegister.register(response);
			} catch (Exception exception) {
				ResponseRegister.register(response, "The request could not be completed.", HttpServletResponse.SC_INTERNAL_SERVER_ERROR, new HashMap<>());
			}
		}
	}

	private UsernamePasswordAuthenticationToken getUsernamePasswordAuthenticationToken(HttpServletRequest request) {
		String username = getUsername(request);
		if (username != null) {
			UserData userData = authorizationUserService.userData(username);
			if (userData != null && userData.userRoles() != null) {
				return getAuthenticationToken(request, userData);
			}
		}
		throw new AccessDeniedException("Access was not authorized on this request.");
	}

	private UsernamePasswordAuthenticationToken getAuthenticationToken(HttpServletRequest request, UserData userData) {
		List<SimpleGrantedAuthority> authorities = Arrays.stream(userData.userRoles())
				.map(role -> new SimpleGrantedAuthority("ROLE_".concat(role)))
				.collect(Collectors.toList());
		User principal = new User(userData.getUsername(), userData.getUsername(), authorities);
		authorizationHandler.authorizationSuccessfully(request, userData);
		return new UsernamePasswordAuthenticationToken(principal, null, principal.getAuthorities());
	}

	private String getUsername(HttpServletRequest request) throws UnsupportedOperationException {
		String authorization = request.getHeader(AUTHORIZATION)
				.replace("Bearer ", "");

		return Jwts.parser()
				.setSigningKey(authorizationUserService.secret())
				.parseClaimsJws(authorization)
				.getBody()
				.getSubject();
	}

	private boolean shouldBeAuthorized(HttpServletRequest request) {
		String requestHeader = request.getHeader(AUTHORIZATION);
		String requestURI = request.getRequestURI();
		return !(requestHeader == null ||
				requestURI.contains(Constants.OPEN_API_URL));
	}
}

File: src/main/kotlin/com/softwareplace/authorization/JWTAuthorizationFilter.kt
package com.softwareplace.authorization

import com.softwareplace.authorization.ResponseRegister.register
import com.softwareplace.config.ApplicationInfo
import com.softwareplace.model.toAuthorizationUser
import com.softwareplace.service.AuthorizationUserService
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.MalformedJwtException
import io.jsonwebtoken.SignatureException
import org.springframework.http.HttpHeaders
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter
import org.springframework.web.client.HttpClientErrorException.Unauthorized
import java.io.IOException
import javax.servlet.FilterChain
import javax.servlet.ServletException
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

open class JWTAuthorizationFilter(
    authenticationManager: AuthenticationManager?,
    private val authorizationUserService: AuthorizationUserService,
    private val authorizationHandler: AuthorizationHandler,
    private val applicationInfo: ApplicationInfo
) : BasicAuthenticationFilter(authenticationManager) {

    @Throws(IOException::class, ServletException::class)
    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, chain: FilterChain) {
        try {
            SecurityContextHolder.getContext().authentication = getUsernamePasswordAuthenticationToken(request)
            chain.doFilter(request, response)
        } catch (exception: Exception) {
            when (exception) {
                is Unauthorized,
                is SignatureException,
                is AccessDeniedException,
                is MalformedJwtException -> {
                    response.status = HttpServletResponse.SC_UNAUTHORIZED
                    register(response)
                }
                else -> register(response, ERROR_RESPONSE_MESSAGE, HttpServletResponse.SC_BAD_REQUEST, HashMap())
            }

            authorizationHandler.onAuthorizationFailed(request, response, chain, exception)
        }
    }

    private fun getUsernamePasswordAuthenticationToken(request: HttpServletRequest): UsernamePasswordAuthenticationToken {
        val authToken = getAuthToken(request)
        val userData = authorizationUserService.findUser(authToken)

        userData?.run {
            authorizationHandler.authorizationSuccessfully(request, userData)
            request.setAttribute(USER_SESSION_DATA, userData)
            val authorizationUser = this.toAuthorizationUser()
            return UsernamePasswordAuthenticationToken(authorizationUser, null, authorizationUser.authorities)
        }

        throw AccessDeniedException(UNAUTHORIZED_ERROR_MESSAGE)
    }

    @Throws(UnsupportedOperationException::class)
    fun getAuthToken(request: HttpServletRequest): String {
        val requestHeader = request.getHeader(HttpHeaders.AUTHORIZATION)

        requestHeader?.run {
            val authorization = this
                .replace(BEARER, "")

            return Jwts.parser()
                .setSigningKey(applicationInfo.securitySecret)
                .parseClaimsJws(authorization)
                .body
                .subject
        }

        throw AccessDeniedException(UNAUTHORIZED_ERROR_MESSAGE)
    }

    companion object {
        const val BEARER = "Bearer "
        const val UNAUTHORIZED_ERROR_MESSAGE = "Access was not authorized on this request."
        const val ROLE = "ROLE_"
        const val ERROR_RESPONSE_MESSAGE = "The request could not be completed."
        const val USER_SESSION_DATA = "USER_SESSION_DATA"
    }
}
